### ISSUE TITLE
TypeError: bwipjs.toBuffer is not a function

### ISSUE THREAD

**Issue by AloneSpace** [metafloor/bwip-js#156]
i used electron + vue.js how to fix this

**Comment by metafloor** [comment]
https://www.youtube.com/watch?v=Gc9ilHp01vY

**Comment by steffanhalv** [comment]
I have this error to. The documentation says .toBuffer is a valid function, but it does not exist.

**Comment by metafloor** [comment]
If you cannot bother to post some code and explain what you are seeing and doing, there is no way to understand the issue.  Have you run the debugger and looked at what the exported bwipjs object has as properties?

**Comment by steffanhalv** [comment]
> n

**Here are the properties from the console log:**

(Missing the function called `toBuffer()`)

![Skjermbilde 2020-05-08 kl  16 26 02](https://user-images.githubusercontent.com/4593184/81415485-ce1e8180-9148-11ea-8499-18566c4dbd1a.png)

And here is the documentation from the docs in readme:

```js
<script>
    var bwipjs = require('bwip-js');
    bwipjs.toBuffer({ bcid:'qrcode', text:'0123456789' }, function (err, png) {
        if (err) {
          document.getElementById('output').textContent = err;
        } else {
          document.getElementById('myimg').src = 'data:image/png;base64,' +
                                                 png.toString('base64');
        }
      });
  </script>
```

**Comment by metafloor** [comment]
It appears the electron packager is importing the browser version.  You need to change your configuration to force electron to use the node version.

**Comment by steffanhalv** [comment]
Ah I see. I ended up using the browser version, as I will use it for both browser and electron. Perhaps the docs in readme could be updated for the electron version.

But thank you for fast response :)

**Comment by metafloor** [comment]
Not sure what changed with electron.  Its packager used to grab the version pointed to by `main` in a module's `package.json`.

**Comment by steffanhalv** [comment]
My electron version is 8.2.2, but it should not be a problem for electron users to use the canvas method anyway as it normaly uses the browser window.

**Comment by Garric15** [comment]
I have the same problem, but I'm using React Typescript. How do I specify to get the node version ? I can see it in the dist folder, but I'm not able to use it.

**Comment by metafloor** [comment]
@Garric15: Are you using react-native with typescript?  Is that why the need for the node version?

If you look at the package.json file in the main bwip-js directory, it has:

      "main": "./dist/node-bwipjs.js",
      "browser": "./dist/bwip-js.js",

I cannot think of a reason why react-native would not pull from `main`.  But I don't use react or react-native, so I may not understand some of its assumptions and requirements.  If some one knows the answer, please let us know!

**Comment by jucgonza** [comment]
Hi, I had a similar problem with Ionic + React + Typescript, using version 2.0.10 of bwip-js, in first place I had to install the Type definitions for bwip-js which come for the 1.7.3 version only, then I got this error and many others when trying to use the plugin: `Property 'toCanvas' does not exist on type 'typeof BwipJs'...
`

I finally solved it by adding the type definition for the property 'toCanvas'  to the definitions file like this:
`...
declare namespace BwipJs {
    export function loadFont(fontName:string, sizeMulti: number, fontFile: string): void;
    export function toCanvas(opts:BwipJs.ToBufferOptions,canvas: string | HTMLCanvasElement): string | HTMLCanvasElement;
    export function toBuffer(opts: ToBufferOptions, callback:(err: string|Error, png: Buffer) => void): void;
    interface ToBufferOptions {
        bcid: string;
...`

Hopefully this solution could help someone.

**Comment by metafloor** [comment]
I do not use typescript (yet).  Is there a file of type definitions for bwip-js that should be included with the project?  And are there any "linkages" that need to be setup in `package.json` or in the source code to refer to the type defs?

**Comment by jucgonza** [comment]
Well, I'm not completely sure how typescript works, but I had to install the type definitions for bwipjs in order to be able to use it with React as typescript suggested. I installed it using `npm install @types/bwip-js` I don't think it should be included with the project since it is maintained by other people in this repository: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/bwip-js

As for the additional setup in package.json, there is no additional setup for the type defs as far as I know

But as you mentioned above I had to edit  package.json file in the main bwip-js directory as follows:
`"browser": "./dist/node-bwipjs.js",`

Then toBuffer function worked perfectly.

**Comment by marco-visualsquares** [comment]
Face same problem for vue2 + typescript. So I don't think it is related to electron.

**Comment by Garric15** [comment]
I think by using typescript we want the `main` version. Here is what the current package.json look like:

```json
"main": "./dist/node-bwipjs.js",
"browser": "./dist/bwip-js.js",
```

Webpack actually decides on what to load. And because it's a web app (e.g. React, Vue), it will look first at the `browser` field, and if it doesn't exist, it'll look for the `module`, and lastly `main`.

It explain why the function is not available, the browser version is loaded. But I still don't know how to fix this.

Although it work by editing the package.json like @jucgonza suggested, I don't think it's a great idea for collaborative projects, as we should not change anything in the node_modules folder.

**Comment by metafloor** [comment]
@marco-visualsquares : You mentioned that you are using vue.  Are you expecting to have node's `toBuffer()` interface in the browser?

**Comment by marco-visualsquares** [comment]
@metafloor Thx for the quick reply! Well, I am using vue-cli + typescript to build my project, which actually node-based, so it would be reasonable to expect I can access node's `toBuffer()` interface.

**Comment by metafloor** [comment]
Where does the code execute?  In node or on the browser?  Doesn't matter how you build/bundle the code, only where the final code executes.  If you are using vue, then I suspect the code executes in the browser.

**Comment by marco-visualsquares** [comment]
Yes I am running in browser eventually. Is there any work-around/alternatives? `toCanvas()` is not suitable for me as the canvas element is not always there.

**Comment by metafloor** [comment]
You can always create a canvas element as needed e.g. `let canvas = document.createElement('canvas')`. It doesn't have to be added to the DOM.
